presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotSegmentPageSourceForGrpc.java

```diff
@@ -0,0 +1,142 @@
+            long startTimeNanos = System.nanoTime();
+            ServerResponseWrapper serverResponse = serverResponseIterator.next();
+            readTimeNanos += System.nanoTime() - startTimeNanos;
+            switch (serverResponse.getResponseType()) {
+                case Constants.Response.ResponseType.DATA:
+                    estimatedMemoryUsageInBytes = serverResponse.getSerializedSize();
+                    // Store each dataTable which will later be constructed into Pages.
+                    try {
+                        byteBuffer = serverResponse.getPayloadReadOnlyByteBuffer();
+                        DataTable dataTable = serverResponse.getDataTable(byteBuffer);
+                        checkExceptions(dataTable, split, PinotSessionProperties.isMarkDataFetchExceptionsAsRetriable(session));
+                        currentDataTable = new PinotSegmentPageSource.PinotDataTableWithSize(dataTable, serverResponse.getSerializedSize());
+                    }
+                    catch (IOException e) {
+                        throw new PinotException(
+                            PINOT_DATA_FETCH_EXCEPTION,
+                            split.getSegmentPql(),
+                            String.format("Encountered Pinot exceptions when fetching data table from %s: %s", split, e));
+                    }
+                    break;
+                case Constants.Response.ResponseType.METADATA:
+                    // The last part of the response is Metadata
+                    currentDataTable = null;
+                    serverResponseIterator = null;
+                    close();
+                    return null;
+                default:
+                    throw new PinotException(
+                        PINOT_UNEXPECTED_RESPONSE,
+                        split.getSegmentPql(),
+                        String.format("Encountered Pinot exceptions, unknown response type - %s", serverResponse.getResponseType()));
+            }
+        }
+        Page page = fillPage();
+        completedPositions += currentDataTable.getDataTable().getNumberOfRows();
+        if (byteBuffer != null) {
+            byteBuffer.clear();
+        }
+        return page;
+    }
+
+    private Iterator<ServerResponseWrapper> queryPinot(PinotSplit split)
+    {
+        String sql = split.getSegmentPql().orElseThrow(() -> new PinotException(PINOT_INVALID_PQL_GENERATED, Optional.empty(), "Expected the segment split to contain the pql"));
+        String host = split.getSegmentHost().orElseThrow(() -> new PinotException(PINOT_INVALID_PQL_GENERATED, Optional.empty(), "Expected the segment split to contain the host"));
+        String[] hostSplits = host.split("_");
+        String grpcHost = (hostSplits.length > 1) ? hostSplits[hostSplits.length - 2] : host;
+        int grpcPort = split.getGrpcPort().isPresent() ? split.getGrpcPort().get() : 0;
+        // Fallback to default grpc port set by Pinot config.
```

Comment:
This is more about system integration and upgrade steps:

Pinot added the gRpcPort field into instance config later than the streaming connector, so this API call may not get the real gRpc port back, hence we need a *fallback* global gRpc port.

Basically the logic is that, if we can read the port from instance config, we use it, otherwise go to look at the default gRpc port.